scripts/extensions/videoEditor/src/VideoTimeline/VideoTimeline.tsx

```diff
@@ -0,0 +1,294 @@
+        // Don't use event timeupdate, because cannot set timeloop, default is 250ms
+        this.intervalTimer = window.setInterval(this.tick, 100);
+        // Use dragover event window to save position X, because event.clientX can not get in Firefox
+        document.addEventListener('dragover', this.handleDragOver);
+        this.setThumbnailsRender();
+    }
+
+    componentDidUpdate(prevProps: IProps) {
+        if (prevProps.thumbnails !== this.props.thumbnails) {
+            this.setThumbnailsRender();
+        }
+        if (this.props.trim !== prevProps.trim) {
+            this.updateTrim(this.props.trim.start, this.props.trim.end);
+        }
+    }
+
+    componentWillUnmount() {
+        clearInterval(this.intervalTimer);
+        document.removeEventListener('dragover', this.handleDragOver);
+    }
+
+    updateTrim(start: number, end: number) {
+        this.setState({
+            trim: {
+                start: start,
+                end: end,
+            },
+        });
+    }
+
+    handleDrag(type: 'left' | 'right') {
+        let time = this.getPositionInBar(this.positionX) * this.props.video.duration;
+
+        if (type === 'left' && time < this.state.trim.end - VIDEO_MIN_DURATION) {
+            this.updateTrim(time, this.state.trim.end);
+        }
+        if (type === 'right' && time > this.state.trim.start + VIDEO_MIN_DURATION) {
+            this.updateTrim(this.state.trim.start, time);
+        }
+    }
+
+    handleDragOver(e: DragEvent | React.DragEvent<HTMLDivElement>) {
+        if (e.clientX) {
+            this.positionX = e.clientX;
+        }
+    }
+
+    // select a number of thumbnails will be rendered based on screen width
+    // because there is not enough space for all of them
+    setThumbnailsRender() {
+        if (this.props.thumbnails.length === 0) {
+            this.setState({thumbnailsRender: []});
+            return;
+        }
+
+        const width = this.props.thumbnails[0].width;
+        // Plus one more incomplete frame because we do not know thumbnails width beforehand to
+        // make timeline render with just enough width space
+        const total = Math.floor((this.controlbar.current?.offsetWidth ?? 1) / width) + 1;
+        const secondPerFrame = (this.props.video.duration - 1) / (this.props.thumbnails.length - 1);
+        const startPosition = width / 2 + (this.controlbar.current?.getBoundingClientRect()?.x ?? 0);
+
+        const thumbnailsIndex = Array.from(
+            Array(total).keys(),
+            (i) => {
+                const videoTime = this.getPositionInBar(startPosition + i * width) * this.props.video.duration;
+
+                return Math.round(videoTime / secondPerFrame);
+            },
+        );
+
+        let thumbnails = this.props.thumbnails.filter((_, index) => thumbnailsIndex.includes(index));
+
+        // missing some thumbnails on the last second of video, happen when video is too short (<15s)
```

Comment:
Can you fix this on the back-end so the thumbnaIls aren't missing?

Reply:
We exclude last second because `ffmpeg` (editor tools behind `video-server`) could not capture image on end frames (and 1 second is not much), we are testing with lower value (e.g. 0.05) to see if it's worked and will update this.

Reply:
Even if `ffmpeg` can't generate one thumbnail for example and you want to repeat the last one twice, you should do it on the back-end.

Reply:
That's also an option, we will check which side will be easier and update to you